### DISCUSSION TITLE
Can I use psycopg2 and psycopg3 in same project?

### DISCUSSION THREAD

**Discussion by purushottamnawale** [psycopg/psycopg#895] [Q&A] [answered]
Can I use psycopg2 and psycopg3 in same project? Will it cause any issues if I used it different python files in django?

**Comment by AumGupta** [comment]
**No, it's generally recommended to use a single library (either psycopg2 or psycopg3) consistently throughout your Django project.** Switching between them in different files can lead to compatibility issues.

While both libraries are for interacting with PostgreSQL databases, they have different APIs and incompatible codebases. Using them together would likely lead to conflicts and errors.

**Choose one library and stick with it.**

Here's a simple example using psycopg2:

```python
import psycopg2

conn = psycopg2.connect(
    dbname="your_database_name",
    user="your_username",
    password="your_password",
    host="your_host",
    port="your_port"
)

cursor = conn.cursor()
cursor.execute("SELECT * FROM your_table")
rows = cursor.fetchall()

for row in rows:
    print(row)

cursor.close()
conn.close()
```

If you still need to use both (I would not recommend), make sure to clearly separate their usage in your code to avoid confusion.

---
If it helped, let me know. Happy coding!

**Reply by purushottamnawale** [reply]
Most of our codebase is written in psycopg2. But I wanted to use some new features of psycopg3 in new files/code without changing existing code. So, I wanted to know will it cause any conflicts or unexpected behaviour?

**Reply by AumGupta** [reply]
See it is not recommended as they **can** conflict. If you still wanna go with it then my friend, nothing is impossible.
> You can try a *workaround*.
> 1. **Create a separate module.** Isolate psycopg3 usage in a new module or project within your application. This will prevent direct conflicts between psycopg2 and psycopg3 imports and API usage.
> 2. **Use a dependency injection framework:** Implement a dependency injection framework like `injector` or `python-di` to manage dependencies and inject the appropriate library instance into your code. This allows you to dynamically choose between psycopg2 and psycopg3 based on configuration or context.

I'd still say that maintaining it in future would be even tougher. So if possible try updating you existing codebase and then build on the newer psycopg3. Rest it's your choice.

**Comment by dvarrazzo** [comment, marked as answer]
Yes, psycopg2 and 3 can coexist in the same project and they won't cause each other problems.

As suggested, of course, maintaining a codebase with two subtly different modules doing the same thing might be tricky and that's up to you to manage. I think your use case is clear: you want to leave the bulk of your problem in psycopg2, for now, and use 3 for new components. I see no problem doing so.
